docs/design/daemon.md

```diff
@@ -0,0 +1,128 @@
+
+For other uses, see the related [feature request](https://github.com/GoogleCloudPlatform/kubernetes/issues/1518)
+
+## Functionality
+
+The Daemon Controller supports standard API features:
+- create
+  - The spec for daemon controllers has a pod template field.
+  - Using the pod’s node selector field, Daemon controllers can be restricted to operate over nodes that have a certain label. For example, suppose that in a cluster some nodes are labeled ‘database’. You can use a daemon controller to launch a datastore pod on exactly those nodes labeled ‘database’.
+  - Using the pod's node name field, Daemon controllers can be restricted to operate on a specified node.
+  - The spec for pod templates that run with the Daemon Controller is the same as the spec for pod templates that run with the Replication Controller, except there will not be a ‘replicas’ field (exactly 1 daemon pod will be launched per node).
+  - We will not guarantee that daemon pods show up on nodes before regular pods - run ordering is out of scope for this controller. 
+  - The initial implementation of Daemon Controller does not guarantee that Daemon pods show up on nodes (for example because of resource limitations of the node), but makes a best effort to launch Daemon pods (like Replication Controllers do with pods). Subsequent revisions might ensure that Daemon pods show up on nodes, pushing out other pods if necessary.
+  - A daemon controller named “foo” adds a “controller: foo” annotation to all the pods that it creates
+  - YAML example:
+```YAML
+  apiVersion: v1
+  kind: Daemon
+  metadata:
+    labels:
+      name: datastore
+    name: datastore
+  spec:
+    template:
+      metadata:
+        labels:
+          name: datastore-shard
+      spec:
+        node-selector: 
+          name: datastore-node
+        containers:
+          name: datastore-shard
+          image: kubernetes/sharded
+          ports:
+            - containerPort: 9042
+              name: main
+```
+  - commands that get info
+    - get (e.g. kubectl get dc)
+    - describe
+  - Modifiers
+    - delete
+    - stop: first we turn down all the pods controller by the daemon (by setting the nodeName to a non-existed name). Then we turn down the daemon controller.
+    - label
+    - update
+    - Daemon controllers have labels, so you could, for example, list all daemon controllers with a certain label (the same way you would for a Replication Controller).
+  - In general, for all the supported features like get, describe, update, etc, the Daemon Controller works in a similar way to the Replication Controller. However, note that the Daemon Controller and the Replication Controller are different constructs.
+
```

Comment:
We'd probably describe this in deployment terminology as MaxAvailability:
80% and MaxSurge: 0.  The deployment process would need a way to make the
daemon controller "inert" or "inactive", create a new daemon controller,
and then selectively delete old pods.

On Wed, Sep 2, 2015 at 10:12 PM, Kamal Marhubi <notifications@github.com>
wrote:

> In docs/design/daemon.md
> <https://github.com/kubernetes/kubernetes/pull/13368#discussion_r38607531>
> :
>
> > +          image: kubernetes/sharded
> > +          ports:
> > +            - containerPort: 9042
> > +              name: main
> > +```
> > +  - commands that get info
> > +    - get (e.g. kubectl get dc)
> > +    - describe
> > +  - Modifiers
> > +    - delete
> > +    - stop: first we turn down all the pods controller by the daemon (by setting the nodeName to a non-existed name). Then we turn down the daemon controller.
> > +    - label
> > +    - update
> > +    - Daemon controllers have labels, so you could, for example, list all daemon controllers with a certain label (the same way you would for a Replication Controller).
> > +  - In general, for all the supported features like get, describe, update, etc, the Daemon Controller works in a similar way to the Replication Controller. However, note that the Daemon Controller and the Replication Controller are different constructs.
> > +
>
> I think that the update needs to be explained a bit more carefully. Will
> there be a rolling update to limit the number of instances that are
> changing? The description of the daemon controller's operation doesn't seem
> to allow for that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13368/files#r38607531>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift